pkg/mutation/path/parser/parser.go

```diff
@@ -0,0 +1,163 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package parser
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/open-policy-agent/gatekeeper/pkg/mutation/path/token"
+)
+
+type Parser struct {
+	input     string
+	scanner   *token.Scanner
+	curToken  token.Token
+	peekToken token.Token
+	err       error
+}
+
+// Parse parses the provided input and returns an abstract representation if successful.
+func Parse(input string) (*Path, error) {
+	p := newParser(input)
+	return p.Parse()
+}
+
+func newParser(input string) *Parser {
+	s := token.NewScanner(input)
+	p := &Parser{
+		input:   input,
+		scanner: s,
+	}
+	p.curToken = p.scanner.Next()
+	p.peekToken = p.scanner.Next()
+	return p
+}
+
+// next advances to the next token in the stream.
+func (p *Parser) next() {
+	p.curToken = p.peekToken
+	p.peekToken = p.scanner.Next()
+}
+
+// expect returns whether the next token matches our expectation,
+// and if so advances to that token.
+// Otherwise returns false and doesn't advance.
+func (p *Parser) expect(t token.Type) bool {
+	if p.peekToken.Type == t {
+		p.next()
+		return true
+	}
+	return false
+}
+
+// expectPeek returns whether the next token matches our expectation.
+// The current token is not advanced either way.
+func (p *Parser) expectPeek(t token.Type) bool {
+	return p.peekToken.Type == t
+}
+
+func (p *Parser) Parse() (*Path, error) {
```

Comment:
nit: this can be private?

Reply:
I'll keep the method exported in case the parser is exported in the future. It is the public-facing interface to the parser, even if the parser is not public.